### PR TITLE
Non-default MLP weight initializations

### PR DESCRIPTION
**Pull request by amburger66**

### CODE REVIEW

File: src/ml_models.py [hunk 1]
```diff
@@ -334,24 +339,56 @@ def _create_optimizer(self) -> optim.Optimizer:
...
+            optimizer = self._create_optimizer()
+            # Run training.
+            best_loss = _train_pytorch_model(
+                self,
+                loss_fn,
+                optimizer,
+                batch_generator,
+                max_iters=self._max_train_iters,
+                n_iter_no_change=self._n_iter_no_change)
+            if best_loss < 1:
```
Comment (tomsilver) [inline src/ml_models.py:387]: what's the rationale for using 1 here?

File: src/ml_models.py [hunk 2]
```diff
@@ -334,24 +339,56 @@ def _create_optimizer(self) -> optim.Optimizer:
...
-                             loss_fn,
-                             optimizer,
-                             batch_generator,
-                             max_iters=self._max_train_iters,
-                             n_iter_no_change=self._n_iter_no_change)
+        for _ in range(self._n_reinitialize_tries):
+            # (Re-)initialize weights.
+            self._reset_weights()
+            # Create the loss function.
+            loss_fn = self._create_loss_fn()
```
Comment (tomsilver) [inline src/ml_models.py:376]: does the loss function need to be created multiple times?

File: src/ml_models.py [hunk 3]
```diff
@@ -334,24 +339,56 @@ def _create_optimizer(self) -> optim.Optimizer:
...
+            # Create the optimizer.
+            optimizer = self._create_optimizer()
```
Comment (tomsilver) [inline src/ml_models.py:378]: same question about the optimizer, although _maybe_ it has some kind of internal state

**Comment by amburger66** [comment]
coverenv_weight_init_glib_num_transitions_perc_solved
coverenv_weight_init_glib_num_transitions_query_cost

**Comment by amburger66** [comment]
seed 456:

https://user-images.githubusercontent.com/26547725/164950031-11a733fa-f054-4be1-91f6-c215d2aca59b.mp4

https://user-images.githubusercontent.com/26547725/164950033-6c37bbe2-607b-49e8-b050-5101dcfdf772.mp4

**Comment by amburger66** [comment]
coverenv_weight_init_greedy_100k_iters_num_transitions_query_cost

coverenv_weight_init_greedy_100k_iters_num_transitions_perc_solved

**Comment by ronuchit** [comment]
what's the latest with this PR? are we looking to merge it?
